File: claude_atlas/vaults.py
"""Create new claude-obsidian vaults and register vaults as tree leaves."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from .console import Console
from .errors import AtlasError
from .home import Config, display_path, tree_root
from .product import ClaudeObsidian, now_utc, operation_id
from .tree import Node, create_leaf, find_by_vault, slugify, walk


def resolve_new_vault_path(arg: str, vaults_dir: Path) -> Path:
    """A bare name lands in the vaults directory; anything path-like is a path."""
    if "/" in arg or arg.startswith(("~", ".")):
        return Path(arg).expanduser().resolve()
    return (vaults_dir / arg).resolve()


def create_vault(
    product: ClaudeObsidian, path: Path, console: Console, *, confirm: bool = True
) -> dict[str, Any]:
    """Run claude-obsidian's plan-then-apply init as one reviewed step."""
    if path.exists():
        raise AtlasError(
            f"{display_path(path)} already exists; use `claude-atlas vault add` "
            "to register an existing vault"
        )
    path.parent.mkdir(parents=True, exist_ok=True)
    generated_at = now_utc()
    operation = operation_id("init", generated_at)
    plan = product.init_plan(path, generated_at=generated_at, operation=operation)
    if plan.get("status") != "dry-run":
        raise AtlasError(f"unexpected init plan status {plan.get('status')!r}")
    changed = plan.get("changed_paths", [])
    if confirm:
        console.say(f"claude-obsidian will create {display_path(path)} with {len(changed)} files:")
        for item in changed:
            console.say(f"    {item}")
        console.say()
        if not console.confirm("Create this vault?"):
            raise AtlasError("cancelled")
    return product.init_apply(
        path,
        generated_at=generated_at,
        operation=operation,
        approval=str(plan["approved_plan_sha256"]),
    )


def register_vault(
    config: Config,
    vault: Path,
    *,
    name: str | None = None,
    parent: str | None = None,
    purpose: str = "",
    priority: str = "normal",
) -> Node:
    vault = vault.expanduser().resolve()
    if not (vault / ".claude-obsidian.json").is_file():
        raise AtlasError(
            f"{display_path(vault)} is not a claude-obsidian vault "
            "(no .claude-obsidian.json)"
        )
    root = tree_root(config)
    existing = find_by_vault(walk(root), vault)
    if existing is not None:
        raise AtlasError(f"{display_path(vault)} is already registered as {existing.rel}")
    display = name or vault.name
    directory = create_leaf(
        root,
        node_id=slugify(display),
        name=display,
        vault=vault,
        parent=parent,
        purpose=purpose,
        priority=priority,
    )
    from .tree import load_node

    return load_node(directory, root)

File: internal/vaults/vaults.go
// Package vaults creates and adopts vaults and registers them as project pages in the tree.
package vaults

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/nathanaday/claude-atlas/internal/console"
	"github.com/nathanaday/claude-atlas/internal/home"
	"github.com/nathanaday/claude-atlas/internal/tree"
	"github.com/nathanaday/claude-atlas/internal/vault"
)

var ErrCancelled = errors.New("cancelled")

// ResolveNewPath puts a bare name under the vaults directory; anything path-like is a path.
func ResolveNewPath(arg, vaultsDir string) (string, error) {
	if strings.Contains(arg, string(filepath.Separator)) || strings.HasPrefix(arg, "~") || strings.HasPrefix(arg, ".") {
		return filepath.Abs(home.Expand(arg))
	}
	return filepath.Abs(filepath.Join(vaultsDir, arg))
}

// Create makes a new vault at path after showing what it will contain.
func Create(path string, mode vault.Mode, c *console.Console, confirm bool) (*vault.InitResult, error) {
	if _, err := os.Stat(path); err == nil {
		return nil, fmt.Errorf("%s already exists; use `claude-atlas adopt` for an existing vault", home.Display(path))
	}
	if mode == "" {
		mode = vault.Generic
	}
	if confirm {
		files := append(vault.TemplateFiles(), vault.Marker, vault.LedgerPath)
		c.Say("claude-atlas will create %s (%s mode) with %d files and a git repository:", home.Display(path), mode, len(files))
		for _, item := range files {
			c.Say("    %s", item)
		}
		c.Say("")
		ok, err := c.Confirm("Create this vault?", true)
		if err != nil {
			return nil, err
		}
		if !ok {
			return nil, ErrCancelled
		}
	}
	return vault.Init(path, mode, time.Now())
}

// RegisterOptions are the authored fields for a new project.
type RegisterOptions struct {
	Name     string
	Category string
	Purpose  string
	Priority string
}

// Register adds a project page pointing at an existing vault.
func Register(cfg *home.Config, root string, opts RegisterOptions) (*tree.Project, error) {
	root, err := filepath.Abs(home.Expand(root))
	if err != nil {
		return nil, err
	}
	if !vault.IsVault(root) && !vault.IsLegacy(root) {
		return nil, fmt.Errorf("%s is not a claude-atlas vault (no %s); create one with `claude-atlas new-vault` or adopt it with `claude-atlas adopt`", home.Display(root), vault.Marker)
	}
	treeRoot := cfg.TreeRoot()
	projects, _, err := tree.Walk(treeRoot)
	if err != nil {
		return nil, err
	}
	if existing := tree.FindByVault(projects, root); existing != nil {
		return nil, fmt.Errorf("%s is already registered as %s", home.Display(root), existing.Rel)
	}
	name := opts.Name
	if name == "" {
		name = filepath.Base(root)
	}
	id, err := tree.Slugify(name)
	if err != nil {
		return nil, err
	}
	path, err := tree.Create(treeRoot, tree.ProjectOptions{
		ID: id, Name: name, Vault: root, Category: opts.Category, Purpose: opts.Purpose, Priority: opts.Priority,
	})
	if err != nil {
		return nil, err
	}
	return tree.Load(path, treeRoot)
}
